harness/tests/cli/test_auth.py

```diff
@@ -29,74 +34,502 @@
+        # ---
+        # Explicit user but no password, DET_USER and DET_PASS set, and DET_USER_TOKEN unset.
+        # DET_PASS still ignored since DET_USER/DET_PASS are meant to be processed as a unit, and
+        # an explicitly-requested username overrides that unit.
+        Login("y", "n", "y", "y", "n", "y", CheckCache, UseCache),
+        Login("y", "n", "y", "y", "n", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "y", "y", "n", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # Explicit user but no password, DET_USER set but no other env.  DET_USER is ignored.
+        Login("y", "n", "y", "n", "n", "y", CheckCache, UseCache),
+        Login("y", "n", "y", "n", "n", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "y", "n", "n", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # DET_USER_TOKEN is overridden by a configured cache (the on-cluster `det user login` case).
+        Login("*", "*", "ye", "*", "y", "y", CheckCache, UseCache),
+        # DET_USER_TOKEN can be used if user is explicit, so long as DET_USER matches explicit user.
+        Login("y", "*", "y", "*", "y", "n", Use("user", "env_token")),
+        # No explicit user; token in env is used if cache is missing or invalid.
+        Login("n", "*", "y", "*", "y", "n", Use("user", "env_token")),
+        Login("n", "*", "y", "*", "y", "x", CheckExpired, Use("user,", "env_token")),
+        # No explicit user; token in env is overridden by a configured cache.
+        # (the on-cluster `det user login` case).
+        Login("n", "*", "ye", "*", "y", "y", CheckCache, UseCache),
+        # ---
+        # Nothing explicit; DET_USER and DET_PASS are set, DET_USER_TOKEN is unset.
+        # Cache continues to work where it matches DET_USER, and there are no password prompts.
+        Login("n", "*", "y", "y", "n", "y", CheckCache, UseCache),
+        Login("n", "*", "y", "y", "n", "n", DoLogin("user", "env_pass"), UseNew),
+        Login("n", "*", "y", "y", "n", "x", CheckExpired, DoLogin("user", "env_pass"), UseNew),
+        # ---
+        # Nothing explicit; and DET_USER is ignored without either DET_PASS or DET_USER_TOKEN.
+        Login("n", "n", "*", "n", "n", "n", DoLogin("determined", ""), UseNew),
+        # the username is taken from the cache but password must be provided again
+        Login("n", "n", "*", "n", "n", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        Login("n", "n", "*", "n", "n", "y", CheckCache, UseCache),
+        # ---
+        # If password is explicit but username is not, we fall back to the default username.
+        Login("n", "y", "n", "n", "n", "n", DoLogin("determined", "req_pass"), UseNew),
+        # Other pass-but-not-user cases are governed by other effects (see several earlier cases).
+        Login("n", "y", "*", "*", "*", "y", CheckCache, UseCache),
+    ],
```

Comment:
these test cases are hard to grok. can't easily tell which cases have been covered, which aren't, etc. i wonder if organizing them by functional globs would help? like, if it's possible to isolate some of the parameters with their expected function calls. i.e. `Login("*", "*", "*", "*", "*", "x", CheckExpired)` 

i guess downside to this would be creating lots of duplicate "scenarios" and running through duplicate asserts. but this could be avoided if we globbed in the other direction. like, we could parametrize all the parameters, loop through each permutation individually, then match each one to expected function calls?

i dunno, i guess the actual login code is just way too complicated for these tests to be clean.